lib/pbench/server/api/resources/quisby.py

```diff
@@ -0,0 +1,104 @@
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    APIInternalError,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+    Parameter,
+    ParamType,
+    Schema,
+)
+from pbench.server.cache_manager import (
+    CacheManager,
+    TarballNotFound,
+    TarballUnpackError,
+)
+from pbench.server.database import Dataset
+
+
+class QuisbyData(ApiBase):
+    """
+    API class to retrieve Quisby data for a dataset
+    """
+
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True),
+                ),
+                authorization=ApiAuthorizationType.DATASET,
+            ),
+        )
+
+    def _get(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        This function returns the Quisby data for the requested dataset.
+
+        Args:
+            params: includes the uri parameters, which provide the dataset.
+            request: Original incoming Request object
+            context: API context dictionary
+
+        Raises:
+            APIAbort, reporting "NOT_FOUND" and "INTERNAL_SERVER_ERROR"
+
+        GET /api/v1/quisby/{dataset}
+        """
+
+        dataset = params.uri["dataset"]
+
+        cache_m = CacheManager(self.config, current_app.logger)
+        try:
+            tarball = cache_m.find_dataset(dataset.resource_id)
+        except TarballNotFound as e:
+            raise APIAbort(HTTPStatus.NOT_FOUND, str(e))
+
+        name = Dataset.stem(tarball.tarball_path)
+        try:
+            file = tarball.extract(tarball.tarball_path, f"{name}/result.csv")
+        except TarballUnpackError as e:
+            raise APIInternalError(str(e)) from e
+
+        metadata = self._get_dataset_metadata(
+            dataset, ["dataset.metalog.pbench.script"]
+        )
+        benchmark = metadata["dataset.metalog.pbench.script"]
+        if benchmark == "uperf":
+            benchmark_type = BenchmarkName.UPERF
+        elif benchmark == "fio":
+            benchmark_type = BenchmarkName.FIO
+        else:
+            raise APIAbort(HTTPStatus.NOT_FOUND, "Unsupported Benchmark")
+
+        get_quisby_data = QuisbyProcessing.extract_data(
+            self, benchmark_type, dataset.name, InputType.STREAM, file
+        )
+
+        if get_quisby_data["status"] == "success":
+            response = jsonify(get_quisby_data)
+            response.status_code = HTTPStatus.OK
+            return response
+        else:
+            current_app.logger.error(
+                "Quisby processing failure. Exception :{}", get_quisby_data["exception"]
+            )
+            raise APIAbort(
+                HTTPStatus.INTERNAL_SERVER_ERROR, "Unexpected failure from Quisby"
+            )
```

Comment:
Never use `APIAbort` for internal errors. Unfortunately it's really easy. (Maybe I should even catch that with an assert in the constructor, though that's really too late.) Use `APIInternalError` instead, which has additional semantics.

Use the `logger.error` message above as the message parameter: that will be put into the journal along with a UUID, but the user will get a short "internal error" message along with the UUID so that the associated explanation can be found in the journal.

Reply:
> Maybe I should even catch that with an assert in the constructor, though that's really too late.

Yes, I think you should:  it wouldn't be too late when the code is run during development and testing, right?  And, arguably, when we run in production, we should be turning assertions off, and so it wouldn't have any impact then.